File: apps/app-form/src/intent.ts
/**
 * @module intent
 * @description Two-tier intent detection for form interactions
 *
 * ## Design Rationale
 *
 * Intent detection determines what the user wants to do:
 * - Fill form fields? Submit? Cancel? Undo?
 *
 * We use a two-tier approach:
 *
 * ### Tier 1: Fast Path (English Keywords)
 *
 * Simple regex matching for common English phrases. This is:
 * - **Fast**: No LLM call, instant response
 * - **Reliable**: Simple patterns rarely fail
 * - **Limited**: Only works for English, explicit commands
 *
 * Used for: "submit", "undo", "cancel", "skip", etc.
 *
 * ### Tier 2: LLM Fallback
 *
 * When fast path returns null, we use LLM for:
 * - **Non-English**: "enviar" (Spanish for "submit")
 * - **Ambiguous**: "I think I'm done?" (submit or just commenting?)
 * - **Complex**: Multi-part messages with intent + data
 *
 * The LLM call also extracts field values, so we bundle
 * intent detection with extraction to save latency.
 *
 * ## Why Fast Path First
 *
 * 1. **Latency**: LLM calls take 500ms-2s. Fast path is <1ms.
 * 2. **Cost**: Each LLM call has token cost. Fast path is free.
 * 3. **Reliability**: Regex patterns are deterministic. LLM can be unpredictable.
 *
 * ## Intent Categories
 *
 * **Lifecycle Intents** - Change session state:
 * - submit, stash, restore, cancel
 *
 * **UX Intents** - Helper actions:
 * - undo, skip, explain, example, progress, autofill
 *
 * **Data Intents** - Provide information:
 * - fill_form (with extractions)
 *
 * **Fallback**:
 * - other (unknown intent, may still have data)
 */

import type { FormIntent } from "./types";

/**
 * Quick intent detection using English keywords.
 *
 * Fast path for common English phrases - avoids LLM call for obvious intents.
 * Returns null if no match found, triggering LLM fallback.
 *
 * WHY regex-based:
 * - Deterministic - same input always gives same output
 * - Fast - no API calls, no parsing
 * - Easy to add patterns
 *
 * WHY word boundaries (\b):
 * - Prevents false matches ("undoing" matching "undo")
 * - Allows matching in context ("I want to cancel")
 *
 * @param text - User message text
 * @returns Detected intent or null if no fast-path match
 */
export function quickIntentDetect(text: string): FormIntent | null {
  const lower = text.toLowerCase().trim();

  // Empty or too short - probably not a command
  // WHY: Single letters or empty strings aren't meaningful commands
  if (lower.length < 2) {
    return null;
  }

  // ═══ LIFECYCLE INTENTS ═══
  // These change the session state

  // Restore - resume a saved form
  // WHY these phrases: Natural ways to continue interrupted work
  if (/\b(resume|continue|pick up where|go back to|get back to)\b/.test(lower)) {
    return "restore";
  }

  // Submit - complete and submit the form
  // WHY "all set" included: Common natural confirmation phrase
  if (/\b(submit|done|finish|send it|that'?s all|i'?m done|complete|all set)\b/.test(lower)) {
    return "submit";
  }

  // Stash - save for later without submitting
  // WHY exclude "save and submit": User wants to submit, not stash
  if (/\b(save|stash|later|hold on|pause|save for later|come back|save this)\b/.test(lower)) {
    // But not "save" in context of submitting
    if (!/\b(save and submit|save and send)\b/.test(lower)) {
      return "stash";
    }
  }

  // Cancel - abandon the form
  // WHY "quit" and "exit": Common ways to abandon a process
  if (/\b(cancel|abort|nevermind|never mind|forget it|stop|quit|exit)\b/.test(lower)) {
    return "cancel";
  }

  // ═══ UX MAGIC INTENTS ═══
  // Helper actions that don't directly provide data

  // Undo - revert the last change
  // WHY "oops" and "that's wrong": Natural error expressions
  if (/\b(undo|go back|wait no|change that|oops|that'?s wrong|wrong|not right)\b/.test(lower)) {
    return "undo";
  }

  // Skip - skip the current optional field
  // WHY exclude "skip to": That's navigation, not skipping current field
  if (/\b(skip|pass|don'?t know|next one|next|don'?t have|no idea)\b/.test(lower)) {
    // But not "skip to" (navigation)
    if (!/\bskip to\b/.test(lower)) {
      return "skip";
    }
  }

  // Explain - "why do you need this?"
  // WHY end anchor: "why" at end of sentence is usually a question
  if (
    /\b(why|what'?s that for|explain|what do you mean|what is|purpose|reason)\b\??$/i.test(lower)
  ) {
    return "explain";
  }
  // Also match standalone "why?"
  if (/^why\??$/i.test(lower)) {
    return "explain";
  }

  // Example - "give me an example"
  // WHY end anchor: "like what?" is asking for example
  if (/\b(example|like what|show me|such as|for instance|sample)\b\??$/i.test(lower)) {
    return "example";
  }
  // Also match standalone "example?" or "e.g.?"
  if (/^(example|e\.?g\.?)\??$/i.test(lower)) {
    return "example";
  }

  // Progress - "how far am I?"
  // WHY these phrases: Natural ways to ask about completion status
  if (/\b(how far|how many left|progress|status|how much more|where are we)\b/.test(lower)) {
    return "progress";
  }

  // Autofill - "use my usual values"
  // WHY: Users with repeat forms want saved values
  if (/\b(same as|last time|use my usual|like before|previous|from before)\b/.test(lower)) {
    return "autofill";
  }

  // No match - let LLM handle it
  // WHY null return: Signals caller to use LLM fallback
  return null;
}

// ============================================================================
// INTENT HELPERS
// ============================================================================

/**
 * Check if intent is a lifecycle intent (affects session state).
 *
 * Lifecycle intents:
 * - submit: Completes the form
 * - stash: Saves for later
 * - restore: Resumes a saved form
 * - cancel: Abandons the form
 *
 * WHY this helper:
 * - Lifecycle intents need special handling
 * - Often preempt normal message processing
 * - May need confirmation (e.g., cancel)
 */
export function isLifecycleIntent(intent: FormIntent): boolean {
  return ["submit", "stash", "restore", "cancel"].includes(intent);
}

/**
 * Check if intent is a UX intent (doesn't directly provide data).
 *
 * UX intents:
 * - undo: Revert last change
 * - skip: Skip optional field
 * - explain: "Why do you need this?"
 * - example: "Give me an example"
 * - progress: "How far am I?"
 * - autofill: "Use my saved values"
 *
 * WHY this helper:
 * - UX intents are helper actions
 * - Don't extract data, just modify session or provide info
 * - Agent response is informational
 */
export function isUXIntent(intent: FormIntent): boolean {
  return ["undo", "skip", "explain", "example", "progress", "autofill"].includes(intent);
}

/**
 * Check if intent likely contains data to extract.
 *
 * Data intents:
 * - fill_form: User is providing field values
 * - other: Unknown intent, may have data
 *
 * WHY this helper:
 * - Determines if we should run extraction
 * - fill_form and other may have inline data
 * - Lifecycle and UX intents don't have data
 */
export function hasDataToExtract(intent: FormIntent): boolean {
  return intent === "fill_form" || intent === "other";
}

File: packages/python/elizaos/advanced_capabilities/form/intent.py
"""Two-tier intent detection for form interactions.

Tier 1: Fast path -- English keyword matching via regex (instant, free).
Tier 2: LLM fallback -- for non-English, ambiguous, or complex messages.
"""

from __future__ import annotations

import re
from typing import TYPE_CHECKING

from .types import FormIntent

if TYPE_CHECKING:
    pass

# ---------------------------------------------------------------------------
# Fast-path patterns
# ---------------------------------------------------------------------------

_SUBMIT_RE = re.compile(r"\b(submit|done|finish|complete|send|confirm)\b", re.IGNORECASE)
_STASH_RE = re.compile(
    r"\b(stash|save for later|pause|put on hold|come back later)\b", re.IGNORECASE
)
_RESTORE_RE = re.compile(r"\b(restore|resume|continue|pick up where|get back to)\b", re.IGNORECASE)
_CANCEL_RE = re.compile(r"\b(cancel|abort|stop|quit|nevermind|never mind)\b", re.IGNORECASE)
_UNDO_RE = re.compile(r"\b(undo|go back|revert|take that back|oops)\b", re.IGNORECASE)
_SKIP_RE = re.compile(r"\b(skip|pass|next|don't need|not now)\b", re.IGNORECASE)
_EXPLAIN_RE = re.compile(
    r"\b(explain|why do you need|what is this for|help me understand)\b",
    re.IGNORECASE,
)
_EXAMPLE_RE = re.compile(r"\b(example|sample|show me|what should|give me an)\b", re.IGNORECASE)
_PROGRESS_RE = re.compile(r"\b(progress|how far|how much left|status|where am i)\b", re.IGNORECASE)
_AUTOFILL_RE = re.compile(
    r"\b(autofill|use (my )?saved|prefill|fill from|use last)\b", re.IGNORECASE
)


def quick_intent_detect(text: str) -> FormIntent | None:
    """Quick intent detection using English keywords.

    Returns the detected intent or ``None`` if no fast-path match.
    """
    text = text.strip()
    if not text:
        return None

    if _SUBMIT_RE.search(text):
        return "submit"
    if _STASH_RE.search(text):
        return "stash"
    if _RESTORE_RE.search(text):
        return "restore"
    if _CANCEL_RE.search(text):
        return "cancel"
    if _UNDO_RE.search(text):
        return "undo"
    if _SKIP_RE.search(text):
        return "skip"
    if _EXPLAIN_RE.search(text):
        return "explain"
    if _EXAMPLE_RE.search(text):
        return "example"
    if _PROGRESS_RE.search(text):
        return "progress"
    if _AUTOFILL_RE.search(text):
        return "autofill"

    return None


# ---------------------------------------------------------------------------
# Intent classification helpers
# ---------------------------------------------------------------------------

_LIFECYCLE_INTENTS: frozenset[FormIntent] = frozenset({"submit", "stash", "restore", "cancel"})
_UX_INTENTS: frozenset[FormIntent] = frozenset(
    {"undo", "skip", "explain", "example", "progress", "autofill"}
)
_DATA_INTENTS: frozenset[FormIntent] = frozenset({"fill_form", "other"})


def is_lifecycle_intent(intent: FormIntent) -> bool:
    """Return ``True`` if the intent changes session state."""
    return intent in _LIFECYCLE_INTENTS


def is_ux_intent(intent: FormIntent) -> bool:
    """Return ``True`` if the intent is a helper action."""
    return intent in _UX_INTENTS


def has_data_to_extract(intent: FormIntent) -> bool:
    """Return ``True`` if the intent likely contains extractable data."""
    return intent in _DATA_INTENTS
